### PR TITLE
switch encoding of Binary NodeId Like Strings to Base64

### PR DESCRIPTION
**Pull request by treestructure**
the aim of this Pull Request is to implement the proposed binary encoding formats of nodeIds. The current version of the OPC-UA spec uses Base64 encoding:

https://reference.opcfoundation.org/v104/Core/docs/Part6/5.3.1/#Table20

The examples given indicate that Base64 is used for opaque / bytestring nodeIds. Also the encoding of Bytestring NodeId Strings in XML files are clearly specified as base64. Also most other SDKs use base64 as well so I would suggest to also use base64

### CODE REVIEW

File: packages/node-opcua-nodeid/test/test_nodeid.js [hunk 1]
```diff
@@ -45,13 +45,13 @@ describe("testing NodeIds", function() {
 
     it("should create a OPAQUE nodeID", function() {
         const buffer = Buffer.alloc(4);
-        buffer.writeUInt32BE(0xdeadbeef, 0);
+        buffer.write("deadbeef", "base64");
```
Comment (erossignon) [inline packages/node-opcua-nodeid/test/test_nodeid.js]: buffer.writeUInt32BE(0xdeadbeef, 0) was intended . please restore

**Review by erossignon — CHANGES_REQUESTED** [review]
Hi @treestructure, 
Can you explain the idea behind switch the representation of a stringified opaque nodeId from "hex" encoding to base64 encoding ?
This change will be a breaking change.

File: packages/node-opcua-server/source/opcua_server.ts [hunk 2]
```diff
@@ -280,7 +280,7 @@ function _serverEndpointsForCreateSessionResponse(server: OPCUAServer, endpointU
     return server
         ._get_endpoints(endpointUrl)
         .map(getRequiredEndpointInfo)
-        .filter((e) => matchUri(e.endpointUrl, endpointUrl));
+        .filter((e) => matchUri(e.endpointUrl, endpointUrl) && !(e as any).restricted);
```
Comment (erossignon) [inline packages/node-opcua-server/source/opcua_server.ts]: The title of the pull request is misleading.. 
Can you move this change into a separate PR and add a associated unit test ?

File: packages/node-opcua-nodeid/test/test_nodeid.js (hunk 1 above)
Comment (treestructure) [reply packages/node-opcua-nodeid/test/test_nodeid.js]: @erossignon - here comes the explanation for this pull request:

the aim of this Pull Request is to implement the proposed binary encoding formats of nodeIds. The current version of the OPC-UA spec uses Base64 encoding:

https://reference.opcfoundation.org/v104/Core/docs/Part6/5.3.1/#Table20

The examples given indicate that Base64 is used for opaque / bytestring nodeIds. Also the encoding of Bytestring NodeId Strings in XML files are clearly specified as base64. Also most other SDKs use base64 as well so I would suggest to also use  base64

**Comment by treestructure** [comment]
the aim of this Pull Request is to implement the proposed binary encoding formats of nodeIds. The current version of the OPC-UA spec uses Base64 encoding:

https://reference.opcfoundation.org/v104/Core/docs/Part6/5.3.1/#Table20

The examples given indicate that Base64 is used for opaque / bytestring nodeIds. Also the encoding of Bytestring NodeId Strings in XML files are clearly specified as base64. Also most other SDKs use base64 as well so I would suggest to also use base64

File: packages/node-opcua-nodeid/test/test_nodeid.js (hunk 1 above)
Comment (treestructure) [reply packages/node-opcua-nodeid/test/test_nodeid.js]: > I have adopted the other test cases for binary node string. If this is a breaking change, is there a requirement or hint to note it somewhere in this repository (Release Notes, Documentation or anything similar) ?  

We would add it to the release note.

**Review by erossignon — COMMENTED** [review]
Please keep the test as is:
```` 
 buffer.writeUInt32BE(0xdeadbeef, 0);
```

File: packages/node-opcua-nodeid/test/test_nodeid.js [hunk 3]
```diff
@@ -45,13 +45,13 @@ describe("testing NodeIds", function() {
...
 
         const nodeId = new NodeId(NodeIdType.BYTESTRING, buffer, 4);
-        nodeId.value.toString("hex").should.equal("deadbeef");
```
Comment (erossignon) [inline packages/node-opcua-nodeid/test/test_nodeid.js:51]: Please restore : nodeId.value.toString("hex").should.equal("deadbeef");

File: packages/node-opcua-nodeid/test/test_nodeid.js [hunk 4]
```diff
@@ -45,13 +45,13 @@ describe("testing NodeIds", function() {
...
+        nodeId.value.toString("base64").should.equal("deadbeef");
         nodeId.namespace.should.equal(4);
         nodeId.identifierType.should.eql(NodeIdType.BYTESTRING);
-        nodeId.toString().should.eql("ns=4;b=deadbeef");
+        nodeId.toString().should.eql("ns=4;b=ZGVhZGJlZWY=");
```
Comment (erossignon) [inline packages/node-opcua-nodeid/test/test_nodeid.js]: this one is OK , as it shows the base64 encoding when nodeId is converted to string form.

File: packages/node-opcua-nodeid/test/test_nodeid.js [hunk 5]
```diff
@@ -163,21 +163,21 @@ describe("testing coerceNodeId", function() {
         buffer.writeUInt32BE(0xb1dedada, 0);
         buffer.writeUInt32BE(0xb0b0abba, 4);
         const nodeId = coerceNodeId(buffer);
-        nodeId.toString().should.eql("ns=0;b=b1dedadab0b0abba");
-        nodeId.value.toString("hex").should.eql("b1dedadab0b0abba");
```
Comment (erossignon) [inline packages/node-opcua-nodeid/test/test_nodeid.js:167]: Please keep :  nodeId.value.toString("hex").should.eql("b1dedadab0b0abba");

File: packages/node-opcua-nodeid/test/test_nodeid.js [hunk 6]
```diff
@@ -163,21 +163,21 @@ describe("testing coerceNodeId", function() {
...
+        nodeId.toString().should.eql("ns=0;b=sd7a2rCwq7o=");
+        nodeId.value.toString("base64").should.eql("sd7a2rCwq7o=");
     });
 
     it("should coerce a OPAQUE buffer in a string ( with namespace ) ", function() {
-        const nodeId = coerceNodeId("ns=0;b=b1dedadab0b0abba");
+        const nodeId = coerceNodeId("ns=0;b=dGVzdDEyMzQ1Cg==");
         nodeId.identifierType.should.eql(NodeIdType.BYTESTRING);
-        nodeId.toString().should.eql("ns=0;b=b1dedadab0b0abba");
-        nodeId.value.toString("hex").should.eql("b1dedadab0b0abba");
```
Comment (erossignon) [inline packages/node-opcua-nodeid/test/test_nodeid.js:174]: Keep the hex form : nodeId.value.toString("hex").should.eql("b1dedadab0b0abba");

**Review by erossignon — CHANGES_REQUESTED** [review]

File: packages/node-opcua-nodeid/test/test_nodeid.js (hunk 3 above)
Comment (treestructure) [reply packages/node-opcua-nodeid/test/test_nodeid.js:51]: reverted

File: packages/node-opcua-nodeid/test/test_nodeid.js (hunk 5 above)
Comment (treestructure) [reply packages/node-opcua-nodeid/test/test_nodeid.js:167]: reverted

File: packages/node-opcua-nodeid/test/test_nodeid.js (hunk 6 above)
Comment (treestructure) [reply packages/node-opcua-nodeid/test/test_nodeid.js:174]: reverted

**Comment by erossignon** [comment]
Hi Martin,
I have reworked the commits into a single one and create this [PR](https://github.com/node-opcua/node-opcua/pull/1016)  I will merge it today .
Thanks
